src/wallet/coinselection.cpp

```diff
@@ -432,3 +448,52 @@ bool SelectionResult::operator<(SelectionResult other) const
     // As this operator is only used in std::min_element, we want the result that has more inputs when waste are equal.
     return *m_waste < *other.m_waste || (*m_waste == *other.m_waste && m_selected_inputs.size() > other.m_selected_inputs.size());
 }
+
+CAmount SelectionResult::GetInputsWeight() const
+{
+    // WU required to store input counter
+    // 1 vbyte for the default case already counted in non_input_size
+    int weight = (GetSizeOfCompactSize(m_selected_inputs.size()) - 1)*4;
+    bool hasWitnessInputs = false;
+
+    for (const auto& coin : m_selected_inputs) {
+        weight += coin.m_input_weight;
+        hasWitnessInputs |= coin.m_isSegwit;
+    }
+
+    // non_input_size counted witness marker and flag as 1vbyte, correcting this now for correct final fee
+    if (hasWitnessInputs) {
+        weight -= 2;
+    } else {
+        weight -= 4;
+
+        // correct for witness stack size counted but not serialized when there is not witness inputs
+        weight -= m_selected_inputs.size();
+    }
+
+    return weight;
+}
+
+CAmount SelectionResult::GetChange(const CoinSelectionParams& coin_selection_params) const
+{
+    if (m_force_no_change) return 0;
+
+    CAmount change = GetSelectedValue() - m_target;
+    // TODO: rename var maybe?
+    if (!m_use_effective) { // opposite of m_subtract_fee_outputs
+        // m_target already includes not_input_fees
+        // but as they are payed by recipients we can add them back to our change
+        const CAmount not_input_fees = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.tx_noinputs_size);
+        change += not_input_fees;
+    } else {
+        const auto size = GetVirtualTransactionSize(GetInputsWeight(), 0, 0);
+        change -= coin_selection_params.m_effective_feerate.GetFee(size);
+    }
+    assert(change >= 0);
+
+    if (change <= coin_selection_params.m_cost_of_change) { // TODO: less or less_or_equal
+        change = 0;
+    }
+
+    return change;
+}
```

Comment:
In 6423998409736be5728bb5826cfa89e1033ed816 "WIP. Return change from SelectionResult"

I feel like all of the size calculation stuff in this function is unnecessary. We don't need `GetChange` to be exact down to the satoshi. Rather it just needs to be close enough to the real change (and essentially be whether there is going to be a change output) and its exact value can be adjusted later in `CreateTransaction` once we know the real size of the transaction. That adjustment doesn't need to account for dust because that should be taken care of here.

It also seems like it is not necessary to calculate the change on the fly like this. Rather the coin selection algorithms should already know what the change is based on their parameters when they selected. While we don't know the fees exactly due to rolling them into the target, we do know how much was selected and what the input fees should be, so we can calculate the change in the `SelectCoins*` functions and just set it in `SelectionResult`.